File: Semester4/Systems for Design and Implementation/mpp-proiect-java-raulfrandes/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepositories/RefereeDBRepository.java
package triathlon.persistence.dbRepositories;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import triathlon.model.Referee;
import triathlon.model.validator.Validator;
import triathlon.persistence.interfaces.IRefereeRepository;
import triathlon.persistence.RepositoryException;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RefereeDBRepository implements IRefereeRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private final Validator<Referee> refereeValidator;

    public RefereeDBRepository(Properties props, Validator<Referee> refereeValidator) {
        logger.info("Initializing RefereeDBRepository with properties: {}", props);
        this.dbUtils = new JdbcUtils(props);
        this.refereeValidator = refereeValidator;
    }

    @Override
    public Referee findByUsername(String username) {
        if (username == null) {
            throw new RepositoryException("Username cannot be null");
        }
        logger.traceEntry("searching referee with username {}", username);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from referees where username = ?")) {
            preparedStatement.setString(1, username);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String password = resultSet.getString("password");
                Referee referee = new Referee(name, username, password);
                referee.setId(id);
                logger.traceExit(referee);
                return referee;
            }
            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Iterable<Referee> findAll() {
        logger.traceEntry();
        Connection connection = dbUtils.getConnection();
        List<Referee> referees = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from referees")) {
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String username = resultSet.getString("username");
                String password = resultSet.getString("password");
                Referee referee = new Referee(name, username, password);
                referee.setId(id);
                referees.add(referee);
            }
            logger.traceExit(referees);
            return referees;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Referee findById(Long entityId) {
        if (entityId == null) {
            throw new RepositoryException("Id cannot be null");
        }
        logger.traceEntry("searching referee with id {}", entityId);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from referees where id = ?")) {
            preparedStatement.setLong(1, entityId);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                String name = resultSet.getString("name");
                String username = resultSet.getString("username");
                String password = resultSet.getString("password");
                Referee referee = new Referee(name, username, password);
                referee.setId(entityId);
                logger.traceExit(referee);
                return referee;
            }
            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Referee save(Referee entity) {
        if (entity == null) {
            throw new RepositoryException("Entity cannot be null");
        }
        refereeValidator.validate(entity);
        logger.traceEntry("saving referee {}", entity);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection
                .prepareStatement("insert into referees (name, username, password) values (?, ?, ?)",
                        Statement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setString(2, entity.getUsername());
            preparedStatement.setString(3, entity.getPassword());
            int result = preparedStatement.executeUpdate();
            logger.trace("saved {} instances", result);

            ResultSet resultSet = preparedStatement.getGeneratedKeys();
            if (resultSet.next()) {
                Long id = resultSet.getLong(1);
                entity.setId(id);
                logger.traceExit(entity);
                return entity;
            }

            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Referee deleteById(Long entityId) {
        return null;
    }

    @Override
    public Referee update(Referee entity) {
        return null;
    }
}

File: Semester4/Systems for Design and Implementation/mpp-proiect-csharp-raulfrandes/persistence/dbRepositories/RefereeDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using log4net;
using model;
using model.validator;
using persistence.interfaces;

namespace persistence.dbRepositories;

public class RefereeDbRepository: IRefereeRepository
{
    private static readonly ILog Log = LogManager.GetLogger("RefereeDBRepository");

    private IDictionary<String, string> _props;
    private IValidator<Referee> _refereeValidator;

    public RefereeDbRepository(IDictionary<string, string> props, IValidator<Referee> refereeValidator)
    {
        Log.Info("Creating RefereeDbRepository");
        this._props = props;
        this._refereeValidator = refereeValidator;
    }
    
    public IEnumerable<Referee> FindAll()
    {
        Log.InfoFormat("Entering FindAll");
        IDbConnection connection = DbUtils.GetConnection(_props);

        IList<Referee> referees = new List<Referee>();
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        String name = dataReader.GetString(1);
                        string username = dataReader.GetString(2);
                        String password = dataReader.GetString(3);
                        Referee referee = new Referee(name, username, password);
                        referee.Id = id;
                        referees.Add(referee);
                    }
                }
            }

            Log.InfoFormat("Exiting FindAll with value {0}", referees);
            return referees;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Referee? FindById(long entityId)
    {
        Log.InfoFormat("Entering FindById with value {0}", entityId);
        IDbConnection connection = DbUtils.GetConnection(_props);

        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entityId;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        String name = dataReader.GetString(1);
                        string username = dataReader.GetString(2);
                        String password = dataReader.GetString(3); 
                        Referee referee = new Referee(name, username, password);
                        referee.Id = id;
                        Log.InfoFormat("Exiting FindByUsername with value {0}", referee);
                        return referee;
                    }
                    Log.InfoFormat("Exiting FindByUsername with value {0}", null);
                    return null;
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Referee? Save(Referee entity)
    {
        _refereeValidator.Validate(entity);
        Log.InfoFormat("Entering Save with value {0}", entity);
        IDbConnection connection = DbUtils.GetConnection(_props);
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "insert into referees (name, username, password) values (@name, @username, @password)";

                IDbDataParameter paramName = command.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = entity.Name;
                command.Parameters.Add(paramName);

                IDbDataParameter paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.Username;
                command.Parameters.Add(paramUsername);

                IDbDataParameter paramPassword = command.CreateParameter();
                paramPassword.ParameterName = "password";
                paramPassword.Value = entity.Password;
                command.Parameters.Add(paramPassword);

                var rows = command.ExecuteNonQuery();
                Log.InfoFormat("saved {0} instances", rows);

                if (rows > 0)
                {
                    command.CommandText = "select last_insert_rowid()";
                    var result = command.ExecuteScalar();
                    if (result != null && result != DBNull.Value)
                    {
                        long id = Convert.ToInt64(result);
                        entity.Id = id;
                        Log.InfoFormat("Exiting Save with value {0}", entity);
                        return entity;
                    }
                }
            }
            Log.InfoFormat("Exiting Save with value {0}", null);
            return null;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Referee DeleteById(long entityId)
    {
        throw new NotImplementedException();
    }

    public Referee Update(Referee entity)
    {
        throw new NotImplementedException();
    }

    public Referee? FindByUsername(string username)
    {
        Log.InfoFormat("Entering FindByUsername with value {0}", username);
        IDbConnection connection = DbUtils.GetConnection(_props);

        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees where username = @username";
                IDbDataParameter paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = username;
                command.Parameters.Add(paramUsername);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        String name = dataReader.GetString(1);
                        String password = dataReader.GetString(3);
                        Referee referee = new Referee(name, username, password);
                        referee.Id = id;
                        Log.InfoFormat("Exiting FindByUsername with value {0}", referee);
                        return referee;
                    }
                    Log.InfoFormat("Exiting FindByUsername with value {0}", null);
                    return null;
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }
}